master/internal/db/postgres_tasks.go

```diff
@@ -287,22 +237,71 @@ func (db *PgDB) CloseOpenAllocations(exclude []model.AllocationID) error {
 		for _, v := range exclude {
 			excludeStr = append(excludeStr, v.String())
 		}
-
 		excludedFilter = strings.Join(excludeStr, ",")
 	}
 
-	if _, err := db.sql.Exec(`
-	UPDATE allocations
-	SET end_time = greatest(cluster_heartbeat, start_time), state = 'TERMINATED'
-	FROM cluster_id
-	WHERE end_time IS NULL AND
-	($1 = '' OR allocation_id NOT IN (
-		SELECT unnest(string_to_array($1, ','))))`, excludedFilter); err != nil {
+	if _, err := Bun().NewUpdate().Table("allocations", "cluster_id").
+		Set("end_time = greatest(cluster_id.cluster_heartbeat, allocations.start_time)").
+		Set("state = 'TERMINATED'").
+		Where("end_time IS NULL AND (? = '' OR allocation_id NOT IN (SELECT unnest(string_to_array(?, ','))))",
+			excludedFilter, excludedFilter).Exec(ctx); err != nil {
 		return errors.Wrap(err, "closing old allocations")
 	}
 	return nil
 }
 
+// RecordTaskStats record stats for tasks.
+func RecordTaskStats(ctx context.Context, stats *model.TaskStats) error {
+	return RecordTaskStatsBun(ctx, stats)
+}
+
+// RecordTaskStatsBun record stats for tasks with bun.
+func RecordTaskStatsBun(ctx context.Context, stats *model.TaskStats) error {
+	_, err := Bun().NewInsert().Model(stats).Exec(context.TODO())
+	return err
+}
+
+// RecordTaskEndStats record end stats for tasks.
+func RecordTaskEndStats(ctx context.Context, stats *model.TaskStats) error {
+	return RecordTaskEndStatsBun(ctx, stats)
+}
+
+// RecordTaskEndStatsBun record end stats for tasks with bun.
+func RecordTaskEndStatsBun(ctx context.Context, stats *model.TaskStats) error {
+	query := Bun().NewUpdate().Model(stats).Column("end_time").
+		Where("allocation_id = ?", stats.AllocationID).
+		Where("event_type = ?", stats.EventType).
+		Where("end_time IS NULL")
+	if stats.ContainerID == nil {
+		// Just doing Where("container_id = ?", stats.ContainerID) in the null case
+		// generates WHERE container_id = NULL which doesn't seem to match on null rows.
+		// We don't use this case anywhere currently but this feels like an easy bug to write
+		// without this.
+		query = query.Where("container_id IS NULL")
+	} else {
+		query = query.Where("container_id = ?", stats.ContainerID)
+	}
+
+	if _, err := query.Exec(ctx); err != nil {
+		return fmt.Errorf("recording task end stats %+v: %w", stats, err)
+	}
+
+	return nil
+}
+
+// EndAllTaskStats called at master starts, in case master previously crashed.
+func EndAllTaskStats(ctx context.Context) error {
+	_, err := Bun().NewUpdate().Table("task_stats", "cluster_id", "allocations").
```

Comment:
this query is, to me, not legible in bun. what is an update with three tables? the from is implicit? can we just keep the original query with NewRaw

Reply:
it did take me a while to adjust to this, but it's here in the docs: https://bun.uptrace.dev/guide/query-update.html#fqn
the "FROM" tables are the 2nd arg (and onward) in `Table(...)`